<h2>Titanic - Model Boosting</h2>

<h3>Importing libraries</h3>

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

<h3>Loading training & testing datasets</h3>

In [2]:
df_train = pd.read_csv("./data/train.csv")

In [3]:
df_train

Age      Fare  Pclass_1  Pclass_2  Sex_female  SibSp_0  SibSp_1  \
0   -0.571912  0.677439         1         0           1        0        1   
1   -0.495775 -0.496656         0         0           0        1        0   
2   -1.028731 -0.249570         0         0           0        0        1   
3    2.321278  0.125434         0         1           0        0        1   
4   -0.800322 -0.054568         1         0           1        1        0   
..        ...       ...       ...       ...         ...      ...      ...   
707  0.014272  2.275871         1         0           1        0        1   
708  0.037181 -0.493572         0         0           0        1        0   
709  0.798546 -0.465072         0         0           1        0        1   
710 -0.648048  4.592978         1         0           1        0        0   
711  0.494000 -0.171569         0         0           0        0        1   

     SibSp_2  SibSp_3  SibSp_4  SibSp_5  Parch_0  Parch_1  Parch_2  Parch_3  \
0          0        0        0        0        1        0        0        0   
1          0        0        0        0        1        0        0        0   
2          0        0        0        0        0        1        0        0   
3          0        0        0        0        0        1        0        0   
4          0        0        0        0        1        0        0        0   
..       ...      ...      ...      ...      ...      ...      ...      ...   
707        0        0        0        0        1        0        0        0   
708        0        0        0        0        1        0        0        0   
709        0        0        0        0        1        0        0        0   
710        1        0        0        0        0        0        1        0   
711        0        0        0        0        0        1        0        0   

     Parch_4  Parch_5  Embarked_C  Embarked_Q  Survived  
0          0        0           0           0         1  
1          0        0           0           0         0  
2          0        0           0           0         0  
3          0        0           0           0         0  
4          0        0           0           0         1  
..       ...      ...         ...         ...       ...  
707        0        0           1           0         1  
708        0        0           0           0         0  
709        0        0           0           0         0  
710        0        0           1           0         1  
711        0        0           0           0         0  

[712 rows x 20 columns]

In [4]:
X_train = df_train.iloc[:, df_train.columns != "Survived"].values
y_train = df_train.loc[:, "Survived"].values

In [5]:
X_train

array([[-0.57191187,  0.67743918,  1.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.49577531, -0.49665643,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [-1.02873125, -0.24956999,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [ 0.79854628, -0.46507212,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.64804843,  4.59297793,  1.        , ...,  0.        ,
         1.        ,  0.        ],
       [ 0.49400002, -0.17156922,  0.        , ...,  0.        ,
         0.        ,  0.        ]])

In [6]:
y_train

array([1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0,

<h3>K-Neighbors Classifier</h3>

In [7]:
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier()

param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9, 11, 13, 15],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan', 'minkowski']
}
grid_search_knn = GridSearchCV(knn, param_grid_knn, cv=5, scoring='accuracy', verbose=2)
grid_search_knn.fit(X_train, y_train)

Fitting 5 folds for each of 42 candidates, totalling 210 fits
[CV] END ...metric=euclidean, n_neighbors=3, weights=uniform; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=3, weights=uniform; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=3, weights=uniform; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=3, weights=uniform; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=3, weights=uniform; total time=   0.0s
[CV] END ..metric=euclidean, n_neighbors=3, weights=distance; total time=   0.0s
[CV] END ..metric=euclidean, n_neighbors=3, weights=distance; total time=   0.0s
[CV] END ..metric=euclidean, n_neighbors=3, weights=distance; total time=   0.0s
[CV] END ..metric=euclidean, n_neighbors=3, weights=distance; total time=   0.0s
[CV] END ..metric=euclidean, n_neighbors=3, weights=distance; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=5, weights=uniform; total time=   0.0s
[CV] END ...metric=euclidean, n_neighbors=5, we

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'metric': ['euclidean', 'manhattan', 'minkowski'],
                         'n_neighbors': [3, 5, 7, 9, 11, 13, 15],
                         'weights': ['uniform', 'distance']},
             scoring='accuracy', verbose=2)

In [8]:
print("Best Parameters:", grid_search_knn.best_params_)
print("Best Score:", grid_search_knn.best_score_)

Best Parameters: {'metric': 'manhattan', 'n_neighbors': 11, 'weights': 'uniform'}
Best Score: 0.8047276666994977


<h3>Decision Tree Classifier</h3>

In [9]:
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier()

param_grid_tree = {
    "criterion": ["gini", "entropy"],
    "max_depth": [None, 10, 20, 30, 40, 50],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 3, 4]
}

grid_search_tree = GridSearchCV(tree, param_grid_tree, cv=5, scoring="accuracy", verbose=2)
grid_search_tree.fit(X_train, y_train)

Fitting 5 folds for each of 144 candidates, totalling 720 fits
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=5; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=5; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=5; total time=   0.0s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=5; total time=   0.0s
[CV] END criterion=gini, max

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 10, 20, 30, 40, 50],
                         'min_samples_leaf': [1, 2, 3, 4],
                         'min_samples_split': [2, 5, 10]},
             scoring='accuracy', verbose=2)

In [10]:
print("Best Parameters:", grid_search_tree.best_params_)
print("Best Score:", grid_search_tree.best_score_)

Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 10}
Best Score: 0.8314586821629074


<h3>Support Vector Classifier</h3>

In [11]:
from sklearn.svm import SVC

svc = SVC()

param_grid_svc = {
    'C': [0.1, 1, 10, 100],
    'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
    'kernel': ['linear', 'rbf', 'poly', 'sigmoid']
}

grid_search_svc = GridSearchCV(svc, param_grid_svc, cv=5, scoring="accuracy", verbose=2)
grid_search_svc.fit(X_train, y_train)

Fitting 5 folds for each of 80 candidates, totalling 400 fits
[CV] END ......................C=0.1, gamma=1, kernel=linear; total time=   0.0s
[CV] END ......................C=0.1, gamma=1, kernel=linear; total time=   0.0s
[CV] END ......................C=0.1, gamma=1, kernel=linear; total time=   0.0s
[CV] END ......................C=0.1, gamma=1, kernel=linear; total time=   0.0s
[CV] END ......................C=0.1, gamma=1, kernel=linear; total time=   0.0s
[CV] END .........................C=0.1, gamma=1, kernel=rbf; total time=   0.0s
[CV] END .........................C=0.1, gamma=1, kernel=rbf; total time=   0.0s
[CV] END .........................C=0.1, gamma=1, kernel=rbf; total time=   0.0s
[CV] END .........................C=0.1, gamma=1, kernel=rbf; total time=   0.0s
[CV] END .........................C=0.1, gamma=1, kernel=rbf; total time=   0.0s
[CV] END ........................C=0.1, gamma=1, kernel=poly; total time=   0.0s
[CV] END ........................C=0.1, gamma=1

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['linear', 'rbf', 'poly', 'sigmoid']},
             scoring='accuracy', verbose=2)

In [12]:
print("Best Parameters:", grid_search_svc.best_params_)
print("Best Score:", grid_search_svc.best_score_)

Best Parameters: {'C': 100, 'gamma': 0.01, 'kernel': 'rbf'}
Best Score: 0.8188712695754949


<h3>Conclusion</h3>
After boosting all models, the best one is decision tree classifier with the highest accuracy of 83%.

1. K-Neighbors Classifier (0.8047) (hyperparameters):
    - 'metric': 'manhattan';
    - 'n_neighbors': 11; 
    - 'weights': 'uniform';

2. Decision Tree Classifier (0.8314) (hyperparameters):
    - 'criterion': 'entropy';
    - 'max_depth': 10;
    - 'min_samples_leaf': 1; 
    - 'min_samples_split': 10;

3. Support Vector Classifier (0.8188) (hyperparameters):
    - 'C': 100; 
    - 'gamma': 0.01; 
    - 'kernel': 'rbf';